src/main/java/eu/europeana/metis/sandbox/service/workflow/InternalValidationServiceImpl.java

```diff
@@ -11,54 +11,106 @@
+  private final Map<String, Lock> datasetIdLocksMap = new ConcurrentHashMap<>();
+  private final Period mapEvictionPeriod = Period.ofDays(1);
 
-    public InternalValidationServiceImpl(
-            ValidationExecutionService validator, PatternAnalysisService<Step> patternAnalysisService,
-            ExecutionPointService executionPointService) {
-        this.validator = validator;
-        this.patternAnalysisService = patternAnalysisService;
-        this.executionPointService = executionPointService;
+  public InternalValidationServiceImpl(ValidationExecutionService validator,
+      PatternAnalysisService<Step, ExecutionPoint> patternAnalysisService,
+      ExecutionPointService executionPointService) {
+    this.validator = validator;
+    this.patternAnalysisService = patternAnalysisService;
+    this.executionPointService = executionPointService;
+  }
+
+  @Override
+  public RecordInfo validate(Record recordToValidate) {
+    requireNonNull(recordToValidate, "Record must not be null");
+
+    var content = recordToValidate.getContentInputStream();
+    var validationResult = validator.singleValidation(SCHEMA, null, null, content);
+    if (!validationResult.isSuccess()) {
+      throw new RecordValidationException(validationResult.getMessage(),
+          validationResult.getRecordId(), validationResult.getNodeId());
     }
+    try {
+      LOGGER.info("Pattern analysis acquiring lock record id: {}", recordToValidate.getEuropeanaId());
+      generateAnalysis(recordToValidate.getDatasetId(), recordToValidate.getContent());
 
-    @Override
-    public RecordInfo validate(Record recordToValidate) {
-        requireNonNull(recordToValidate, "Record must not be null");
+    } catch (PatternAnalysisException e) {
+      LOGGER.error("An error occurred while processing pattern analysis with record id {}", recordToValidate.getEuropeanaId());
+    }
+    return new RecordInfo(recordToValidate);
+  }
 
-        var content = recordToValidate.getContentInputStream();
-        var validationResult = validator.singleValidation(SCHEMA, null, null, content);
-        if (!validationResult.isSuccess()) {
-            throw new RecordValidationException(validationResult.getMessage(),
-                    validationResult.getRecordId(), validationResult.getNodeId());
-        }
-        try {
-            generateAnalysis(recordToValidate.getDatasetId(), recordToValidate.getContent());
-        } catch (PatternAnalysisException e) {
-            LOGGER.error("An error occurred while processing pattern analysis with record id {}", recordToValidate.getEuropeanaId());
-        }
-        return new RecordInfo(recordToValidate);
+  private void generateAnalysis(String datasetId, byte[] recordContent) throws PatternAnalysisException {
+    final Lock lock = datasetIdLocksMap.computeIfAbsent(datasetId, s -> new ReentrantLock());
+    final ExecutionPoint executionPoint;
+    try {
+      lock.lock();
+      LOGGER.debug("{} lock, Locked", datasetId);
```

Comment:
I just noticed maybe a more descriptive message: ```Pattern analysis dataset {} lock, locked``` and ```Pattern analysis dataset {} lock, unlocked```

Reply:
yes that's more informative.